mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -271,28 +271,25 @@ public void setupTask(TaskAttemptContext taskAttemptContext) {
     }
 
     @Override
-    public boolean needsTaskCommit(TaskAttemptContext taskAttemptContext) {
-      return true;
+    public boolean needsTaskCommit(TaskAttemptContext context) {
+      // We need to commit if this is the last phase of a MapReduce process
+      return TaskType.REDUCE.equals(context.getTaskAttemptID().getTaskID().getTaskType()) ||
+          context.getJobConf().getNumReduceTasks() == 0;
```

Comment:
Can't this just be `true`? I think that every task using the output committer needs to commit.

Reply:
We can define OutputCommitter on job level only, and AFAIK Hive writes only on the last stage of the job be it a Reducer for a MapReduce job, or a Map for a MapOnly job.

This way we can save IO by not generating and reading those unnecessary files, but depending on assumptions of the way how Hive works.

Your thoughts? Still voting for the more general solution?

Reply:
What I mean is that I don't think Hive calls the output committer unless it is writing to the table. The map phase of a write job should call the committer only if it is a map-only job. And the reduce phase will call the committer if the job has a reduce phase. Because the MR framework takes care of calling `commitTask` for the right phase, we can default this to `true` and it should be identical.

You can probably test this assumption by changing the logic here to throw an exception if it is violated.

Reply:
Some HiveRunner tests that do writes that involves a MR job might help too?

Reply:
I have tested this with adding the jars to the Hive master from the apache repo and running the queries in local mode. I got some exception and started to investigate the source.

The actual query I have used for this had an order by so it had both a Mapper and a Reducer phase:
```
insert into purchases select * from purchases order by id;
```
My debugging revealed that the commitTask was called for Mappers and Reducers as well. The JobRunner infrastructure calls those and it does not have any information about if the task has actually written anything to anywhere or not. I think this is the purpose of the _needsTaskCommit()_ method, so the developer of the OutputCommitter could decide.

@massdosage: Do you have an easy example for HiveRunner write tests?

Thanks,
Peter

Reply:
Thanks for testing it out. If Hive does call this even in map stages, then I agree that we should keep the logic that you have here.

Reply:
@pvary I guess it would look very similar to one of the existing HiveRunner tests [here](https://github.com/apache/iceberg/blob/master/mr/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandlerBaseTest.java) but one could do a hiveShell.execute() with an insert, update, etc statement and then do a query to select the results back? 

I'm not sure whether these examples (https://github.com/klarna/HiveRunner/blob/master/src/test/java/com/klarna/hiverunner/InsertIntoTableIntegrationTest.java and https://github.com/klarna/HiveRunner/blob/master/src/test/java/com/klarna/hiverunner/data/InsertIntoTableTest.java) from HiveRunner itself would trigger the OutputFormat for the insert operations? Presumably so if the table under test created is created using the StorageHandler?

I'd also highly recommend testing this in distributed mode, we found lots of issues with the InputFormat that only manifested themselves then. I can help with this if you want as we have a test environment for this setup already.

Reply:
@massdosage: Created HiveRunner tests. Could you please take a look at them? (TestHiveRunnerWrite.java)

Trying out on a real cluster is on my TODO list, and will definitely do, but if you have a test cluster and some ideas it would be good if you take a stab at it too (in Hungary we say, that "More eyes see more"). I tried with HiveOnLLAP, but the OutputCommitter lifecycle is not yet handled correctly there, so we have to try the MR way.

Reply:
@massdosage, @rdblue: Tested out the stuff on an MR Cluster by running the following queries successfully:
```
insert into test_table values(3,"3"),(4,"4");
insert into test_table select * from test_table;
insert into test_table select * from test_table order by id;
insert into test_table select * from test_table_2;
```

Had Map only jobs and Map-Reduce jobs too. The result was checked by Hive select queries and everything seems ok.

So if Adrian's team did not find any issues I would think that this PR is ready to be pushed.
Your thoughts?

Thanks, Peter